wdl-analysis/src/analyzer.rs

```diff
@@ -0,0 +1,478 @@
+    error: Option<Arc<anyhow::Error>>,
+    /// The diagnostics for the document.
+    diagnostics: Arc<[Diagnostic]>,
+    /// The scope of the analyzed document.
+    scope: Arc<DocumentScope>,
+}
+
+impl AnalysisResult {
+    /// Constructs a new analysis result for the given graph node.
+    pub(crate) fn new(node: &DocumentGraphNode) -> Self {
+        let (root, lines, error) = match node.parse_state() {
+            ParseState::NotParsed => {
+                panic!("cannot create a result for an file that hasn't been parsed")
+            }
+            ParseState::Error(e) => (None, None, Some(e.clone())),
+            ParseState::Parsed { root, lines, .. } => {
+                (Some(root.clone()), Some(lines.clone()), None)
+            }
+        };
+
+        let analysis = node.analysis().expect("analysis not completed");
+
+        Self {
+            uri: node.uri().clone(),
+            root,
+            lines,
+            error,
+            diagnostics: analysis.diagnostics().clone(),
+            scope: analysis.scope().clone(),
+        }
+    }
+
+    /// Gets the URI of the document that was analyzed.
+    pub fn uri(&self) -> &Arc<Url> {
+        &self.uri
+    }
+
+    /// Gets the root node of the document that was analyzed.
+    ///
+    /// Returns `None` if the document was not parsed.
+    pub fn root(&self) -> Option<&GreenNode> {
+        self.root.as_ref()
+    }
+
+    /// Gets the AST document of the result.
+    ///
+    /// Returns `None` if the document was not parsed.
+    pub fn document(&self) -> Option<wdl_ast::Document> {
+        self.root.as_ref().map(|r| {
+            wdl_ast::Document::cast(SyntaxNode::new_root(r.clone())).expect("node should cast")
+        })
+    }
+
+    /// Gets the line index for the document.
+    ///
+    /// The line index may be used to convert line/column positions into byte
+    /// offsets and vice versa.
+    ///
+    /// Returns `None` if the document was not parsed.
+    pub fn lines(&self) -> Option<&LineIndex> {
+        self.lines.as_deref()
+    }
+
+    /// Gets the error if the document could not be read.
+    ///
+    /// Returns `None` if the document was parsed.
+    pub fn error(&self) -> Option<&anyhow::Error> {
+        self.error.as_deref()
+    }
+
+    /// Gets the diagnostics associated with the document.
+    pub fn diagnostics(&self) -> &[Diagnostic] {
+        &self.diagnostics
+    }
+
+    /// Gets the scope of the analyzed document.
+    pub fn scope(&self) -> &Arc<DocumentScope> {
+        &self.scope
+    }
+}
+
+/// Represents a position in a document's source.
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Default)]
+pub struct SourcePosition {
+    /// Line position in a document (zero-based).
+    pub line: u32,
+    /// Character offset on a line in a document (zero-based). The meaning of
+    /// this offset is determined by the position encoding.
+    pub character: u32,
```

Comment:
Is this better as `col`? Or am I misunderstanding.

Reply:
Perhaps this is tied to the LSP specification's language—I didn't check but, now that I'm reading further, it looks to be used there.

Reply:
Indeed, it was chosen to match the LSP type; effectively the same thing.